### ISSUE TITLE
Better render names with an hyphen

### ISSUE THREAD

**Issue by naoliv** [mapnik/mapnik#3885]
Originally from https://github.com/gravitystorm/openstreetmap-carto/issues/3181:

-----
Have this node as an example https://www.openstreetmap.org/node/3474947606 with the name as `São João del-Rei`

From zoom level 15 to 9 we have this:
![15](https://user-images.githubusercontent.com/982338/39016876-135bd022-43f8-11e8-8269-4f6358ac11a4.png)![14](https://user-images.githubusercontent.com/982338/39016875-13387f6e-43f8-11e8-9c58-3881eb3062d5.png)![13](https://user-images.githubusercontent.com/982338/39016874-1311835a-43f8-11e8-9253-0b72d83fe6d6.png)![12](https://user-images.githubusercontent.com/982338/39016872-12eef588-43f8-11e8-9ef8-776271580186.png)![11](https://user-images.githubusercontent.com/982338/39016871-12ccd142-43f8-11e8-85a7-4cb71dfb9e32.png)![10](https://user-images.githubusercontent.com/982338/39016870-12aacdc2-43f8-11e8-9a7c-97e337c1b046.png)![9](https://user-images.githubusercontent.com/982338/39016868-127e652a-43f8-11e8-94b3-32d470482935.png)

The name is being split at the `del-Rei` hyphen, while it most probably should not be (the renderer should consider it as a single word and not as a line-wrapping symbol).

For example, instead having this
```
São João del-
     Rei
```
_(due to the hyphen and the current rendering (above), it could be misinterpreted as `São João delRei`)_

we could have it split like this:
```
São João
del-Rei
```
-----

[The last comment](https://github.com/gravitystorm/openstreetmap-carto/issues/3181#issuecomment-383139684) has a more detailed view of this problem.

Is it possible to do something about this, please?

**Comment by kocio-pl** [comment]
I guess it's a matter of priorities, because hyphens should be possible to break at, just less possible than breaking at whitespaces, for example "Zespół Szkół Hotelarsko-Turystyczno-Gastronomicznych" looks quite good:

![screenshot-2018-4-22 openstreetmap](https://user-images.githubusercontent.com/5439713/39089714-d4aa8cd8-45cd-11e8-86f3-83ccfb119672.png)

https://www.openstreetmap.org/way/154400552#map=18/52.20331/21.04421

**Comment by talaj** [comment]
Thanks for the report.

I would like to try to explain why the break is put after the hyphen.

Look at the following condition:

https://github.com/mapnik/mapnik/blob/d8dbe11fd09f55ba6b59c9e497a5a21335acb15b/src/text/text_layout.cpp#L184-L192

There is distinct code branch for space character, which uses [`icu::BreakIterator`](https://ssl.icu-project.org/apiref/icu4c/classicu_1_1BreakIterator.html) to locate boundaries for word wrap. It is quite complicated thing with various rules for various languages. It also wraps on other characters than the space character. More information is [here](http://userguide.icu-project.org/boundaryanalysis).

But this is just wrapping by linguistic rules, without any aesthetic correction.

If you set `wrap-character` to other than the space character, Mapnik will wrap strictly on that character without involving any linguistic rules.

**Comment by sommerluk** [comment]
Thanks for the details explication.

I would suggest to update the documentation of `text-wrap`-character and `shield-wrap-character` at http://mapnik.org/mapnik-reference/ then. Currently, it is:

```
Use this character instead of a space to wrap long text.

Default: " " (Lines will be wrapped when whitespace is encountered.)
```

I would suggest something like:

```
Use this character to wrap long text.

Default: " " (Lines will be wrapped conforming to the [Unicode Line Break Algorithm](http://www.unicode.org/reports/tr14) by [ICU BreakIterator](http://icu-project.org/apiref/icu4c/classicu_1_1BreakIterator.html).)

If you set wrap-character to other than the space character, Mapnik will wrap strictly on that character without involving any additional rules.
```

**Comment by talaj** [comment]
@sommerluk Good idea. Please, create a pull request in [https://github.com/mapnik/mapnik-reference](https://github.com/mapnik/mapnik-reference/blob/d89bcde0e94aefc45aa78441e27fa3dbd3cb5e72/3.0.20/reference.json#L2097-L2104).
